File: packages/sdk/dotnet/Client/StreamResponse.cs
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace Ihui.AI;

/// <summary>
/// 流式响应(SSE 解析工具)— 将 HTTP 响应流解析为 <see cref="IAsyncEnumerable{JsonElement}"/>。
/// </summary>
/// <remarks>
/// 封装 Server-Sent Events 流式响应,逐行解析 <c>data: {json}</c> 格式。
/// 遇到 <c>data: [DONE]</c> 时结束迭代。
/// <para>
/// 用法:
/// <code>
/// await foreach (var chunk in client.Ai.ChatCompletionsStreamAsync(req))
/// {
///     var delta = chunk.GetProperty("choices")[0]
///         .GetProperty("delta")
///         .GetProperty("content")
///         .GetString();
///     Console.Write(delta);
/// }
/// </code>
/// </para>
/// </remarks>
public static class StreamResponse
{
    /// <summary>从 HTTP 响应流解析 SSE 事件,逐 chunk yield JsonElement。</summary>
    /// <param name="response">HTTP 响应消息(必须已 successful)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>IAsyncEnumerable,逐 chunk yield JsonElement</returns>
    public static async IAsyncEnumerable<JsonElement> ParseSseAsync(
        HttpResponseMessage response,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        using (response)
        {
            var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
            using var reader = new StreamReader(stream, Encoding.UTF8);

            while (!reader.EndOfStream)
            {
                var line = await reader.ReadLineAsync(cancellationToken).ConfigureAwait(false);
                if (line is null) break;

                var trimmed = line.Trim();
                if (trimmed.Length == 0) continue;
                if (!trimmed.StartsWith("data:")) continue;

                var payload = trimmed.Substring(5).Trim();
                if (payload == "[DONE]") break;

                JsonElement element;
                try
                {
                    using var doc = JsonDocument.Parse(payload);
                    element = doc.RootElement.Clone();
                }
                catch
                {
                    continue;
                }

                yield return element;
            }
        }
    }
}

File: packages/sdk/java/src/main/java/com/ihui/ai/sdk/StreamResponse.java
package com.ihui.ai.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import okhttp3.Response;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * 流式响应(SSE 解析迭代器)。
 *
 * <p>封装 Server-Sent Events 流式响应,逐行解析 {@code data: {json}} 格式。
 * 遇到 {@code data: [DONE]} 时结束迭代。
 *
 * <p>用法(try-with-resources):
 * <pre>
 * try (StreamResponse stream = client.ai.completionsStream(req)) {
 *     while (stream.hasNext()) {
 *         JsonNode chunk = stream.next();
 *         String delta = chunk.path("choices").path(0).path("delta").path("content").asText("");
 *         System.out.print(delta);
 *     }
 * }
 * </pre>
 */
public final class StreamResponse implements Iterator<JsonNode>, AutoCloseable {

    private final Response response;
    private final BufferedReader reader;
    private JsonNode nextEvent;
    private boolean closed = false;

    /**
     * 构造 StreamResponse。
     *
     * @param response OkHttp Response(必须已 successful 且有 body)
     * @throws IOException 打开读取流失败
     */
    public StreamResponse(Response response) throws IOException {
        this.response = response;
        this.reader = BaseClient.newReader(response);
        this.nextEvent = parseNext();
    }

    @Override
    public boolean hasNext() {
        return nextEvent != null;
    }

    @Override
    public JsonNode next() {
        if (nextEvent == null) {
            throw new NoSuchElementException("Stream exhausted");
        }
        JsonNode current = nextEvent;
        try {
            nextEvent = parseNext();
        } catch (IOException e) {
            throw new SdkException(0, "stream_read_error",
                    "Stream read failed: " + e.getMessage(), null);
        }
        return current;
    }

    private JsonNode parseNext() throws IOException {
        String line;
        while ((line = reader.readLine()) != null) {
            String trimmed = line.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            if (!trimmed.startsWith("data:")) {
                // 跳过 event:/注释/心跳行
                continue;
            }
            String payload = trimmed.substring(5).trim();
            if ("[DONE]".equals(payload)) {
                return null;
            }
            try {
                return JsonUtil.MAPPER.readTree(payload);
            } catch (IOException e) {
                // 跳过无法解析的行
            }
        }
        return null;
    }

    @Override
    public void close() {
        if (closed) {
            return;
        }
        closed = true;
        try {
            reader.close();
        } catch (IOException e) {
            // 静默关闭
        }
        response.close();
    }
}
